AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+
+
+#import "ASVideoNode.h"
+
+@interface ASVideoNode () {
+  ASDN::RecursiveMutex _lock;
+  
+  __weak id<ASVideoNodeDatasource> _datasource;
+  
+  BOOL _shouldBePlaying;
+  AVAsset *_asset;
+  AVPlayerItem *_currentItem;
+  ASButtonNode *_playButton;
+  ASDisplayNode *_playerNode;
+  ASDisplayNode *_spinner;
+}
+
+@end
+
+@interface ASDisplayNode ()
+- (void)setInterfaceState:(ASInterfaceState)newState;
+@end
+
+@implementation ASVideoNode
+
+- (instancetype)init {
+  if (!(self = [super init])) { return nil; }
+  
+  _playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    playerLayer.player = [[AVPlayer alloc] init];
+    return playerLayer;
+  }];
+  [self addSubnode:_playerNode];
+  
+  self.gravity = ASVideoGravityResizeAspect;
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(restartVideo:) name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
+  
+  return self;
+}
+
+- (void)setInterfaceState:(ASInterfaceState)newState
+{
+  [super setInterfaceState:newState];
+  
+  if (!(newState & ASInterfaceStateVisible)) {
+    [self pause];
+    [_spinner removeFromSupernode];
+  } else {
+    if (_shouldBePlaying) {
+      [self play];
+    }
+    if (_spinner) {
+      [self addSubnode:_spinner];
+    }
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
```

Comment:
Yeah, I'm pretty sure AVPlayerItemStatusUnknown is 0 and AVPlayerItemStatusReadyToPlay is 1